File: Go/Lesson_079_subtree_of_another_tree.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 079 -- Subtree of Another Tree
// Category   : Trees
// Difficulty : Easy
// Study Plan : Day 40
// =============================================================
//
// QUESTION:
//   Given roots of two trees root and subRoot, return true if subRoot is a subtree of root.
// =============================================================

package main
import "fmt"
type TreeNode struct { Val int; Left, Right *TreeNode }
func same(p, q *TreeNode) bool {
    if p == nil && q == nil { return true }
    if p == nil || q == nil || p.Val != q.Val { return false }
    return same(p.Left,q.Left) && same(p.Right,q.Right)
}
func isSubtree(root, sub *TreeNode) bool {
    if root == nil { return false }
    if same(root, sub) { return true }
    return isSubtree(root.Left, sub) || isSubtree(root.Right, sub)
}
func main(){
    r := &TreeNode{Val:3, Left:&TreeNode{Val:4, Left:&TreeNode{Val:1}, Right:&TreeNode{Val:2}}, Right:&TreeNode{Val:5}}
    s := &TreeNode{Val:4, Left:&TreeNode{Val:1}, Right:&TreeNode{Val:2}}
    fmt.Println(isSubtree(r, s))
}

File: Python/Lesson_079_subtree_of_another_tree.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 079 -- Subtree of Another Tree
# Category   : Trees
# Difficulty : Easy
# Study Plan : Day 40
# =============================================================
#
# QUESTION:
#   Given roots of two trees root and subRoot, return true if subRoot is a subtree of root.
# =============================================================

class TreeNode:
    def __init__(self, v=0, l=None, r=None): self.val=v; self.left=l; self.right=r
class Solution:
    def isSameTree(self, p, q):
        if not p and not q: return True
        if not p or not q or p.val != q.val: return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
    def isSubtree(self, root, sub):
        if not root: return False
        if self.isSameTree(root, sub): return True
        return self.isSubtree(root.left, sub) or self.isSubtree(root.right, sub)

if __name__ == "__main__":
    r = TreeNode(3, TreeNode(4, TreeNode(1), TreeNode(2)), TreeNode(5))
    s = TreeNode(4, TreeNode(1), TreeNode(2))
    print(Solution().isSubtree(r, s))
